### ISSUE TITLE
$refs type definition should include `undefined`

### ISSUE THREAD

**Issue by Hoishin** [vuejs/vue#8798]
When you try to refer non-existing ref, it is `undefined`.

```vue
<template>
  <div/>
</template>
<script lang="ts">
import {Vue, Component} from 'vue-property-decorator'

@Component
export default class extends Vue {
  mounted() {
    console.log(this.$refs.foo)
  }
}
```

But the type definition for `$refs` doesn't have undefined, and it is not type safe.

**Comment by JounQin** [comment]
`$refs` could not be `undefined`.
https://github.com/vuejs/vue/blob/master/src/core/instance/lifecycle.js#L39

**Comment by JounQin** [comment]
I don't think we should change it. A lot of things could be `undefined` like properties of `$data`, `$props`, `$options`, etc. And it would be painful to check them everywhere when `strictNullChecks` of `tsconfig.json` is enabled.

Of course, I think it maybe implement somehow in https://github.com/vuejs/eslint-plugin-vue.

**Comment by Hoishin** [comment]
Yes it would be painful. But that does not kill the outcome of type safe code.
TBH if null-check is painful, you shouldn’t use TypeScript.

I will check what’s missing `undefined` and update this issue and the PR. Thanks for letting me know.

**Comment by Hoishin** [comment]
After some discussion in the PR, I would leave this closed and see what version 3.0 will bring.
